crates/trie/common/src/hashed_state.rs

```diff
@@ -515,47 +491,45 @@ impl AsRef<Self> for HashedPostStateSorted {
-    pub fn accounts_sorted(&self) -> impl Iterator<Item = (B256, Option<Account>)> {
-        self.accounts
-            .iter()
-            .map(|(address, account)| (*address, Some(*account)))
-            .chain(self.destroyed_accounts.iter().map(|address| (*address, None)))
-            .sorted_by_key(|entry| *entry.0)
-    }
-
-    /// Extends this collection with contents of another sorted collection.
-    /// Entries in `other` take precedence for duplicate keys.
-    pub fn extend_ref(&mut self, other: &Self) {
-        // Updates take precedence over removals, so we want removals from `other` to only apply to
-        // the previous accounts.
-        self.accounts.retain(|(addr, _)| !other.destroyed_accounts.contains(addr));
-
-        // Extend the sorted accounts vector
-        extend_sorted_vec(&mut self.accounts, &other.accounts);
-
-        // Merge destroyed accounts sets
-        self.destroyed_accounts.extend(&other.destroyed_accounts);
-    }
-}
+// /// Sorted account state optimized for iterating during state trie calculation.
+// #[derive(Clone, Eq, PartialEq, Default, Debug)]
+// pub struct HashedAccountsSorted {
+//     /// Sorted collection of hashed addresses and their account info.
+//     pub accounts: Vec<(B256, Account)>,
+//     /// Set of destroyed account keys.
+//     pub destroyed_accounts: B256Set,
+// }
+
+// impl HashedAccountsSorted {
+//     /// Returns a sorted iterator over updated accounts.
+//     pub fn accounts_sorted(&self) -> impl Iterator<Item = (B256, Option<Account>)> {
+//         self.accounts
+//             .iter()
+//             .map(|(address, account)| (*address, Some(*account)))
+//             .chain(self.destroyed_accounts.iter().map(|address| (*address, None)))
+//             .sorted_by_key(|entry| *entry.0)
+//     }
+//
+//     /// Extends this collection with contents of another sorted collection.
+//     /// Entries in `other` take precedence for duplicate keys.
+//     pub fn extend_ref(&mut self, other: &Self) {
+//         // Updates take precedence over removals, so we want removals from `other` to only apply
+// to         // the previous accounts.
+//         self.accounts.retain(|(addr, _)| !other.destroyed_accounts.contains(addr));
+//
+//         // Extend the sorted accounts vector
+//         extend_sorted_vec(&mut self.accounts, &other.accounts);
+//
+//         // Merge destroyed accounts sets
+//         self.destroyed_accounts.extend(&other.destroyed_accounts);
+//     }
+// }
 
 /// Sorted hashed storage optimized for iterating during state trie calculation.
 #[derive(Clone, Eq, PartialEq, Debug)]
 pub struct HashedStorageSorted {
-    /// Sorted hashed storage slots with non-zero value.
-    pub non_zero_valued_slots: Vec<(B256, U256)>,
-    /// Slots that have been zero valued.
-    pub zero_valued_slots: B256Set,
+    /// Sorted collection of updated storage slots, None indicates zero valued.
+    pub storage_slots: Vec<(B256, Option<U256>)>,
```

Comment:
I believe instead of `Option<U256>` you can use `U256::ZERO` to signal a value removal, unless you discovered some reason this wouldn't be the case.

Reply:
I think we can do it without duplicating significant code, see comment at : https://github.com/paradigmxyz/reth/pull/19233/files#r2481365676